architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,302 @@
+
+Recently, it has been observed that new training techniques are released with an incomplete "preview" version. These "preview" versions tend to be not be fully integrated into OSS. Therefore, using new techniques typically involve additional work. This framework aims to allow timely integrations of such techniques into `sft_trainer.py`. A short list of powerful training techniques but are "preview"-only include:
+- Huggingface integration of [AutoGPTQ](https://github.com/AutoGPTQ/AutoGPTQ).
+    * 4-bit quantization kernels to reduce memory storage of the base weights.
+- [Unsloth](https://github.com/unslothai/unsloth).
+    * Fused operation kernels
+    * Kernels for common model architectures (e.g., cross-entropy losses, RoPE embeddings and RMS norms).
+- [megablocks](https://github.com/databricks/megablocks).
+    * acceleration package for distributing mixture-of-experts that improves upon FSDP sharding.
+
+<!--
+Why this is a valuable problem to solve? What background information is needed to show how this design addresses the problem?
+
+Which users are affected by the problem? Why is it a problem? What data supports this? What related work exists?
+-->
+
+### User Benefit
+
+Users will benefit from powerful training tools integrated into the platform, that are not readily accessible from huggingface. With these tools, users will be able to train models with less GPU resources and/or quicker, resulting in quicker turnaround and improved user experience. 
+
+<!--
+How will users (or other contributors) benefit from this work? What would be the headline in the release notes or blog post?
+-->
+
+## Decision
+
+Terminology | Description
+--|--
+Maintainer | Developers of `sft_trainer.py`.
+Framework | an extensible `Framework` managing all implemented methods.
+Framework Plugin | Self-contained implementation of a framework method.
+
+The proposal satisfies the following desiredata:
+- Unified configuration YAML for all plugins. Fine configuration details abstracted away from Maintainers and other plugin developers.
+- Modular design allows new methods plugins to be added / removed / deactivated seemlessly.
+- Modular design enforces that plugins interact with `sft_trainer.py` at controlled points, and throw appropriate exceptions.
+- Generic enough for most use cases of interest (e.g., quantization, distributed training, etc).
+- Unobstrusive design that only *modifies the model*, and leaves [`SFTTrainer`] unmodified. Minimal inversion-of-control maintained through [`TrainerCallbacks`].
+
+### Only the Model is Modified
+
+Since the [`SFTTrainer`] is designed to work with generic pytorch models, modifying the model is much less intrusive to the training pipeline, then say, modifying [`SFTTrainer`] itself. The hope is then if we constrain ourselves to modify only the model, that we can implement all the method plugins (e.g., quantization, distributed training, etc) that we hope for. 
```

Comment:
> Since the [`SFTTrainer`] is designed to work with generic pytorch models

is this true? it inherits from transformers.Trainer and I see a lot of HF-specific code in Trainer.train().

I know the PEFT code used in SFTTrainer doesn't work nicely with non-HF models

Reply:
Thanks for pointing this out. Ah ok yes `Trainer` is more flexible than `SFTTrainer`; the `SFTTrainer` has some additional requirements (e.g., cannot pass `tokenizer=None`) but these are bare minimum.

I updated the text to be more careful. But you can verify that the below will actually train. This is a ridiculous example, but it does show how `SFTTrainer` can train a `torch.nn.Module`.

Furthermore, if `AccelerationFramework` loads a model, typically it will be one that resembles a HF model. 

```python
iimport numpy as np
import torch

N = 800 # number of examples
S = 10 # sequence length 

# create dataset 
from torch.utils.data.dataset import TensorDataset
rng = np.random.default_rng(1)
data = torch.tensor(rng.random((N, S, 5)), dtype=torch.float)
labels = torch.tensor(rng.random((N, S, 2)), dtype=torch.float)
dataset = TensorDataset(data, labels)

# create data collator
def data_collator(features):
    # import pdb; pdb.set_trace()
    x, y = zip(*features)
    return {'x': torch.stack(x), 'y': torch.stack(y)}

# create model
class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = torch.nn.Linear(5, 2)
        self.loss = torch.nn.MSELoss()
        
    def forward(self, x, y):
        return (self.loss(self.weight(x), y), )

model = Model()

# trainer
from transformers import Trainer
from trl import SFTTrainer
from transformers import TrainingArguments
args = TrainingArguments('results', logging_steps=50, save_strategy='steps', save_steps=50)

# The dataset in this example is overly simple, so just pardon some stubbing of the tokenizer and the
# formatting func
class DummyTokenizer:
    padding_side = None
    save_pretrained = lambda x,y: x
    model_max_length = S

formatting_func = lambda x: x

trainer = SFTTrainer(
    model, args=args, tokenizer=DummyTokenizer(), 
    train_dataset=dataset, data_collator=data_collator,
    formatting_func=lambda x:x,
)
```